In [6]:
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

os.chdir('/Users/karla/SF-dat/SF-DAT-20/Data/') 
data = pd.read_csv('titanic.csv')
data.head()

survived  pclass                                               name  \
0         0       3                            Braund, Mr. Owen Harris   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2         1       3                             Heikkinen, Miss. Laina   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4         0       3                           Allen, Mr. William Henry   

      sex  age  sibsp  parch            ticket     fare cabin embarked  
0    male   22      1      0         A/5 21171   7.2500   NaN        S  
1  female   38      1      0          PC 17599  71.2833   C85        C  
2  female   26      0      0  STON/O2. 3101282   7.9250   NaN        S  
3  female   35      1      0            113803  53.1000  C123        S  
4    male   35      0      0            373450   8.0500   NaN        S

1. How many passengers are in our passenger list? From here forward, we’ll assume our dataset represents the full passenger list for the Titanic.
2. What is the overall survival rate?  ALL ANSWERS AT THE BOTTOM

In [55]:
data.describe()

survived      pclass         age       sibsp       parch        fare  \
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000   
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208   
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429   
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000   
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400   
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200   
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000   
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200   

         IsFemale  
count  891.000000  
mean     0.352413  
std      0.477990  
min      0.000000  
25%      0.000000  
50%      0.000000  
75%      1.000000  
max      1.000000

In [38]:
data['IsFemale'] = data.sex.map({'male':0, 'female':1})
F=data.IsFemale.sum()
print F, data.IsFemale.count()-F
data.groupby(['sex']).describe()



314 577


IsFemale         age        fare       parch      pclass  \
sex                                                                      
female count       314  261.000000  314.000000  314.000000  314.000000   
       mean          1   27.915709   44.479818    0.649682    2.159236   
       std           0   14.110146   57.997698    1.022846    0.857290   
       min           1    0.750000    6.750000    0.000000    1.000000   
       25%           1   18.000000   12.071875    0.000000    1.000000   
       50%           1   27.000000   23.000000    0.000000    2.000000   
       75%           1   37.000000   55.000000    1.000000    3.000000   
       max           1   63.000000  512.329200    6.000000    3.000000   
male   count       577  453.000000  577.000000  577.000000  577.000000   
       mean          0   30.726645   25.523893    0.235702    2.389948   
       std           0   14.678201   43.138263    0.612294    0.813580   
       min           0    0.420000    0.000000    0.000000    1.000000   
       25%           0   21.000000    7.895800    0.000000    2.000000   
       50%           0   29.000000   10.500000    0.000000    3.000000   
       75%           0   39.000000   26.550000    0.000000    3.000000   
       max           0   80.000000  512.329200    5.000000    3.000000   

                   sibsp    survived  
sex                                   
female count  314.000000  314.000000  
       mean     0.694268    0.742038  
       std      1.156520    0.438211  
       min      0.000000    0.000000  
       25%      0.000000    0.000000  
       50%      0.000000    1.000000  
       75%      1.000000    1.000000  
       max      8.000000    1.000000  
male   count  577.000000  577.000000  
       mean     0.429809    0.188908  
       std      1.061811    0.391775  
       min      0.000000    0.000000  
       25%      0.000000    0.000000  
       50%      0.000000    0.000000  
       75%      0.000000    0.000000  
       max      8.000000    1.000000

In [41]:
data.groupby(['survived']).describe()

IsFemale         age        fare       parch      pclass  \
survived                                                                     
0        count  549.000000  424.000000  549.000000  549.000000  549.000000   
         mean     0.147541   30.626179   22.117887    0.329690    2.531876   
         std      0.354968   14.172110   31.388207    0.823166    0.735805   
         min      0.000000    1.000000    0.000000    0.000000    1.000000   
         25%      0.000000   21.000000    7.854200    0.000000    2.000000   
         50%      0.000000   28.000000   10.500000    0.000000    3.000000   
         75%      0.000000   39.000000   26.000000    0.000000    3.000000   
         max      1.000000   74.000000  263.000000    6.000000    3.000000   
1        count  342.000000  290.000000  342.000000  342.000000  342.000000   
         mean     0.681287   28.343690   48.395408    0.464912    1.950292   
         std      0.466660   14.950952   66.596998    0.771712    0.863321   
         min      0.000000    0.420000    0.000000    0.000000    1.000000   
         25%      0.000000   19.000000   12.475000    0.000000    1.000000   
         50%      1.000000   28.000000   26.000000    0.000000    2.000000   
         75%      1.000000   36.000000   57.000000    1.000000    3.000000   
         max      1.000000   80.000000  512.329200    5.000000    3.000000   

                     sibsp  
survived                    
0        count  549.000000  
         mean     0.553734  
         std      1.288399  
         min      0.000000  
         25%      0.000000  
         50%      0.000000  
         75%      1.000000  
         max      8.000000  
1        count  342.000000  
         mean     0.473684  
         std      0.708688  
         min      0.000000  
         25%      0.000000  
         50%      0.000000  
         75%      1.000000  
         max      4.000000

In [51]:
data.groupby(['pclass']).describe()

IsFemale         age        fare       parch       sibsp  \
pclass                                                                     
1      count  216.000000  186.000000  216.000000  216.000000  216.000000   
       mean     0.435185   38.233441   84.154687    0.356481    0.416667   
       std      0.496933   14.802856   78.380373    0.693997    0.611898   
       min      0.000000    0.920000    0.000000    0.000000    0.000000   
       25%      0.000000   27.000000   30.923950    0.000000    0.000000   
       50%      0.000000   37.000000   60.287500    0.000000    0.000000   
       75%      1.000000   49.000000   93.500000    0.000000    1.000000   
       max      1.000000   80.000000  512.329200    4.000000    3.000000   
2      count  184.000000  173.000000  184.000000  184.000000  184.000000   
       mean     0.413043   29.877630   20.662183    0.380435    0.402174   
       std      0.493724   14.001077   13.417399    0.690963    0.601633   
       min      0.000000    0.670000    0.000000    0.000000    0.000000   
       25%      0.000000   23.000000   13.000000    0.000000    0.000000   
       50%      0.000000   29.000000   14.250000    0.000000    0.000000   
       75%      1.000000   36.000000   26.000000    1.000000    1.000000   
       max      1.000000   70.000000   73.500000    3.000000    3.000000   
3      count  491.000000  355.000000  491.000000  491.000000  491.000000   
       mean     0.293279   25.140620   13.675550    0.393075    0.615071   
       std      0.455730   12.495398   11.778142    0.888861    1.374883   
       min      0.000000    0.420000    0.000000    0.000000    0.000000   
       25%      0.000000   18.000000    7.750000    0.000000    0.000000   
       50%      0.000000   24.000000    8.050000    0.000000    0.000000   
       75%      1.000000   32.000000   15.500000    0.000000    1.000000   
       max      1.000000   74.000000   69.550000    6.000000    8.000000   

                survived  
pclass                    
1      count  216.000000  
       mean     0.629630  
       std      0.484026  
       min      0.000000  
       25%      0.000000  
       50%      1.000000  
       75%      1.000000  
       max      1.000000  
2      count  184.000000  
       mean     0.472826  
       std      0.500623  
       min      0.000000  
       25%      0.000000  
       50%      0.000000  
       75%      1.000000  
       max      1.000000  
3      count  491.000000  
       mean     0.242363  
       std      0.428949  
       min      0.000000  
       25%      0.000000  
       50%      0.000000  
       75%      0.000000  
       max      1.000000

In [50]:
data.groupby(['pclass','sex']).describe()

IsFemale         age        fare       parch       sibsp  \
pclass sex                                                                      
1      female count        94   85.000000   94.000000   94.000000   94.000000   
              mean          1   34.611765  106.125798    0.457447    0.553191   
              std           0   13.612052   74.259988    0.728305    0.665865   
              min           1    2.000000   25.929200    0.000000    0.000000   
              25%           1   23.000000   57.244800    0.000000    0.000000   
              50%           1   35.000000   82.664550    0.000000    0.000000   
              75%           1   44.000000  134.500000    1.000000    1.000000   
              max           1   63.000000  512.329200    2.000000    3.000000   
       male   count       122  101.000000  122.000000  122.000000  122.000000   
              mean          0   41.281386   67.226127    0.278689    0.311475   
              std           0   15.139570   77.548021    0.658853    0.546695   
              min           0    0.920000    0.000000    0.000000    0.000000   
              25%           0   30.000000   27.728100    0.000000    0.000000   
              50%           0   40.000000   41.262500    0.000000    0.000000   
              75%           0   51.000000   78.459375    0.000000    1.000000   
              max           0   80.000000  512.329200    4.000000    3.000000   
2      female count        76   74.000000   76.000000   76.000000   76.000000   
              mean          1   28.722973   21.970121    0.605263    0.486842   
              std           0   12.872702   10.891796    0.833930    0.642774   
              min           1    2.000000   10.500000    0.000000    0.000000   
              25%           1   22.250000   13.000000    0.000000    0.000000   
              50%           1   28.000000   22.000000    0.000000    0.000000   
              75%           1   36.000000   26.062500    1.000000    1.000000   
              max           1   57.000000   65.000000    3.000000    3.000000   
       male   count       108   99.000000  108.000000  108.000000  108.000000   
              mean          0   30.740707   19.741782    0.222222    0.342593   
              std           0   14.793894   14.922235    0.517603    0.566380   
              min           0    0.670000    0.000000    0.000000    0.000000   
              25%           0   23.000000   12.331250    0.000000    0.000000   
              50%           0   30.000000   13.000000    0.000000    0.000000   
              75%           0   36.750000   26.000000    0.000000    1.000000   
              max           0   70.000000   73.500000    2.000000    2.000000   
3      female count       144  102.000000  144.000000  144.000000  144.000000   
              mean          1   21.750000   16.118810    0.798611    0.895833   
              std           0   12.729964   11.690314    1.237976    1.531573   
              min           1    0.750000    6.750000    0.000000    0.000000   
              25%           1   14.125000    7.854200    0.000000    0.000000   
              50%           1   21.500000   12.475000    0.000000    0.000000   
              75%           1   29.750000   20.221875    1.000000    1.000000   
              max           1   63.000000   69.550000    6.000000    8.000000   
       male   count       347  253.000000  347.000000  347.000000  347.000000   
              mean          0   26.507589   12.661633    0.224784    0.498559   
              std           0   12.159514   11.681696    0.623404    1.288846   
              min           0    0.420000    0.000000    0.000000    0.000000   
              25%           0   20.000000    7.750000    0.000000    0.000000   
              50%           0   25.000000    7.925000    0.000000    0.000000   
              75%           0   33.000000   10.008300    0.000000    0.000000   
              max           0   74.000000   69.55000

1. How many passengers are in our passenger list? From here forward, we’ll assume our dataset represents the full passenger list for the Titanic.

891 passengers

2. What is the overall survival rate?
38.38% of passengers survived

3. How many male passengers were onboard?
577

4. How many female passengers were onboard? 
314


5. What is the overall survival rate of male passengers?
18.89%


6. What is the overall survival rate of female passengers?
74.2%


7. What is the average age of all passengers onboard? * How did you calculate this average age?
29.69 years old calculated with the describe function.

* Note that some of the passengers do not have an age value. How did you deal with this? What are some other ways of dealing with this?
The describe() function handles a missing value by excluding the observation from the calculation of the mean.
Other ways to deal with the missing values would be to create a dataset and drop the observations where there is no age value, or run a python calculation which takes the sum of only non missing values and divides by the number of items in the sum.  But it seems sensible to me to use a builtin and tested method.
    
  
8. What is the average age of passengers who survived?
28.34

9. What is the average age of passengers who did not survive?
30.26

10. At this (early) point in our analysis, what might you infer about any patterns you are seeing?  
Gender was the biggest influence on survival rate, bigger than age.  More females survived.

11. How many passengers are in each of the three classes of service (e.g. First, Second, Third?)
216, 184, 491

12. What is the survival rate for passengers in each of the three classes of service?
1,2,3 = 63%, 47%, 24%

13. What else might you conclude?
It was better to be in first class. (97% of women survived, 37% of men)
It was better to be female over all. (97% in first, 92% in second, 50% in third) vs (37%, 16%, 14%).
Avoid travelling third class.

A higher percentage of third class was male than in the other two classes. (56%, 59%, 70%).
The survival rate for men in second and third class is very similar(16%, 14%), which is not the case for females (92%,50%).


14. Last, if we were to build a predictive model, which features in the data do you think we should include in the model and which can we leave out? Why?  

I would use gender and class in the predictive model. 

The model might get some fine tuning by knowing whether a person had family on board or not. For instance, what is the survival rate for women who are mothers vs those who are not? This affect might be seen especially in third class, where so many more of the women died than in the other two classes.  And for family men?  The sibsp and parch columns would be useful for this.  

Age is evenly spread between survival groups and within gender groups so not useful for gross prediction purposes, but it might help analyzing family related survival factors.

The cabin numbers are too sparsely recorded to accurately identify traveling companions. I don't think embarkation point is relevant to predicting survival rate.

Fare price reflects class. 

